### ISSUE TITLE
Accept `Output[Resource]` in ~all places where `Resource` is accepted in core SDK

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#7012]
There are a variety of reasons why users may end up with an `Output[Resource]` instead of a `Resource`, and with multi-lang components, this is now also much more common.  However, in several places in the core SDKs, only a `Resource` is accepted, and thus these APIs are hard to use with multi-lang components.

We should accept `Output[Resource]` in all of the following (unless there is a very good reason we can't), in all language SDKs:
* `parent` resource option
* `provider` resource option
* `providers` resource option
* `dependsOn` resource option (already supports `Output<Resource>` in Node.js)

- Go: https://github.com/pulumi/pulumi/pull/7350
- [ ] TypeScript TODO
- [ ] Python TODO
- [ ] .NET TODO

**Comment by benesch** [comment]
For posterity's sake, #5642 has been tracking the `dependsOn` case.

**Comment by t0yv0** [comment]
One implementation issue here is determining child resource set. Currently resource C constructor mutates parent P resource property `__childResources` which gets read in `getAllTransitivelyReferencedCustomResourceURNs` called from `registerResource`, `readResource`. 

In the current implementation it seems possible already that P may complete `registerResource(P)` call from `P.constructor()` before `C.constructor()` registers itself as a child, so P registers with an incomplete child-set. 

If we make parent option an `Input[Resource]` and spawn a thread to add a child to it, we might defer child registration until this Input resolves, which may be even after `C.constructor()` finishes. Moreover it's not clear what should wait for this thread (C and P constructions would be considered complete), so this might introduce races on whether the C-P relation is registered or not.

**Comment by t0yv0** [comment]
This is currently blocked until we have a go-ahead for breaking
changes for these reasons:

- there is no non-breaking path forward for Node JS implementing
  input-ty Parent and Providers deferred options. These are needed for
  transformations and for supporting public interface such as
  [getProvider](https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/resource.ts#L189)

- we cannot ship the feature in Go and have it unimplemented in Node.
  Go [PR](https://github.com/pulumi/pulumi/pull/7411) is ready and
  could be satisfactory with minor revisions

- Fundamentally the reason changes are possible in Go is that the
  runtime provides means to spin-await a deferred value onto the
  caller's stack `Promise<T> -> T`. We have similar facilities in all
  runtimes except Node

- Partial implementations that

- If we can identify a subset of newly added methods that is not
  blocked, I can carve out a subset of the Go PR and have that
  shipped, and follow with Node (and then others)

    # likely not blocked
    func DependsOnInputs(o []ResourceInput) ResourceOption

    # probably blocked (need to attempt a Node implementation to verify)
    func ProviderInput(pri ProviderResourceInput) ResourceOrInvokeOption
    func ProviderInputMap(inputMap map[string]ProviderResourceInput) ResourceOption
    func ProviderInputs(o ...ProviderResourceInput) ResourceOption

    # definitely blocked
    func ParentInput(r ResourceInput) ResourceOrInvokeOption

**Comment by benesch** [comment]
Thanks for the progress update @t0yv0. Just wanted to note that this is (bar none) our biggest pain point with Pulumi in Python, due to the prevalence of multi-lang components. Obviously I'm biased, but the backwards-incompatible changes seem well worth it in exchange for seamless multi-lang components.

**Comment by t0yv0** [comment]
@benesch please upvote the issue (I just did), these are definitely tradeoffs and it may be possible  to reconsider. Also: we do periodically issue breaking changes so the intent here is definitely to fix this, it's a matter of when.

**Comment by lukehoban** [comment]
> there is no non-breaking path forward for Node JS implementing
input-ty Parent and Providers deferred options.

What exactly is the required user-facing breaking change?

**Comment by benesch** [comment]
> > there is no non-breaking path forward for Node JS implementing
> > input-ty Parent and Providers deferred options.
> 
> What exactly is the required user-facing breaking change?

Presumably, at the very least, `Resource.getProvider` needs to return `Output<Provider>` instead of `Provider` directly, since the provider might not be known at the time the `Resource` is constructed?

**Comment by t0yv0** [comment]
Based on our chat with @pgavlin,at the very least, Resource.getProvider becomes async. Possibly (in the worst case), all resource constructions become async and we need to transition from constructors to static methods.  I put in on hold for the moment, but if we're interested I can start building a WIP PR for Node on this, and that would clarify the issues further.

**Comment by mikhailshilkov** [comment]
We need to make a decision very soon whether we are going to fix the relevant part of this issue or if not in the forseeable future, will need to remove this property from pulumi-eks (per https://github.com/pulumi/pulumi-eks/issues/602) since it's unusable.

**Comment by mikhailshilkov** [comment]
Assigning @t0yv0 who wanted to revive this issue and has a lot of the past context. Feel free to engage @RobbieMcKinstry for the parts that are parallelizable.

**Comment by Frassle** [comment]
> (unless there is a very good reason we can't),

We already have a few resource options that we allow to be Outputs today (e.g. alias names). I want to point out that currently these can lead to divergences between what is show in preview and what happens in apply. This will be especially important for using plans as all options must be equal for plans to apply.

The problem is none of these options can be sent to the engine as "unknown". So currently if one of these outputs resolves to unknown at preview time we simply drop it. This can lead to different alias and dependency lists today, and if we extended to allow Output<Resource> for things like parents could lead to resources showing as unparented at preview but then having a parent at update time (this would then lead to completely different URNs).

I doubt this is expected or wanted behavior. For 3.0 we're stuck with the public API we already have and so our choices are:
1) Error if a resource option results in being unknown at preview time
2) Mark the resource as unknown and propagate the unknownness to all it's outputs. This would kinda of match what you get if you try to make a resource inside an output apply which doesn't run.

For 4.0 we could consider making all these options plain, enforcing in the type system that unknown values can't be used for resource options. This would force users to put the new resource call inside an output apply, which would result in a similar end result to option 2 above but a bit more explicitly.

**Comment by mikhailshilkov** [comment]
> For 4.0 we could consider making all these options plain, enforcing in the type system that unknown values can't be used for resource options. This would force users to put the new resource call inside an output apply, which would result in a similar end result to option 2 above but a bit more explicitly.

I think the intent of this issue is the opposite: product-wise there is a need to pass Outputs as options (e.g. a provider that you get from an MLC). Thank you for the caveat though, we should think how we can work around this and make previews and update plans work properly.

**Comment by lukehoban** [comment]
> I wonder if it's more possible to fix that so they aren't outputs? My guess is that we could probably return Resources as Resources because all their properties are outputs anyway. Something to look into. Would be a nicer UX even if we also allowed outputs in settings as per this issue.

Agreed that this would be a great way to resolve this if possible.  I recall there was some sense that it might be possible from @pgavlin - but I also recall a couple of major issues that we weren't sure how to handle.  Pat - any chance you can summarize what makes this hard and what we might be able to do to address?
